In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split 
import time
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
import pickle
import matplotlib.pyplot as plt



In [2]:

def selectkbest(indep_X,dep_Y,n):
        test = SelectKBest(score_func=chi2, k=n)
        fit1= test.fit(indep_X,dep_Y)
        selectk_features = fit1.transform(indep_X)
        return selectk_features
    
def split_scalar(indep_X,dep_Y):
        X_train, X_test, y_train, y_test = train_test_split(indep_X, dep_Y, test_size = 0.25, random_state = 0)
        sc = StandardScaler()
        X_train = sc.fit_transform(X_train)
        X_test = sc.transform(X_test)    
        return X_train, X_test, y_train, y_test
    
 
def cm_prediction(classifier,X_test):
     y_pred = classifier.predict(X_test)
        
        # Making the Confusion Matrix
     from sklearn.metrics import confusion_matrix
     cm = confusion_matrix(y_test, y_pred)
        
     from sklearn.metrics import accuracy_score 
     from sklearn.metrics import classification_report 
        #from sklearn.metrics import confusion_matrix
        #cm = confusion_matrix(y_test, y_pred)
        
     Accuracy=accuracy_score(y_test, y_pred )
        
     report=classification_report(y_test, y_pred)
     return  classifier,Accuracy,report,X_test,y_test,cm

def logistic(X_train,y_train,X_test):       
        # Fitting K-NN to the Training set
        from sklearn.linear_model import LogisticRegression
        classifier = LogisticRegression(random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm      
    
def svm_linear(X_train,y_train,X_test):
                
        from sklearn.svm import SVC
        classifier = SVC(kernel = 'linear', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm
    
def svm_NL(X_train,y_train,X_test):
                
        from sklearn.svm import SVC
        classifier = SVC(kernel = 'rbf', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm
   
def Navie(X_train,y_train,X_test):       
        # Fitting K-NN to the Training set
        from sklearn.naive_bayes import GaussianNB
        classifier = GaussianNB()
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm         
    
    
def knn(X_train,y_train,X_test):
           
        # Fitting K-NN to the Training set
        from sklearn.neighbors import KNeighborsClassifier
        classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm
def Decision(X_train,y_train,X_test):
        
        # Fitting K-NN to the Training set
        from sklearn.tree import DecisionTreeClassifier
        classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm      


def random(X_train,y_train,X_test):
        
        # Fitting K-NN to the Training set
        from sklearn.ensemble import RandomForestClassifier
        classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm
    
def selectk_Classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf): 
    
    dataframe=pd.DataFrame(index=['ChiSquare'],columns=['Logistic','SVMl','SVMnl','KNN','Navie','Decision','Random'])
    for number,idex in enumerate(dataframe.index):      
        dataframe['Logistic'][idex]=acclog[number]       
        dataframe['SVMl'][idex]=accsvml[number]
        dataframe['SVMnl'][idex]=accsvmnl[number]
        dataframe['KNN'][idex]=accknn[number]
        dataframe['Navie'][idex]=accnav[number]
        dataframe['Decision'][idex]=accdes[number]
        dataframe['Random'][idex]=accrf[number]
    return dataframe
    

In [3]:
dataset1=pd.read_csv("Preprocessed_Heart_data.csv",index_col=None)


In [4]:
dataset1

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
0    40.0      140.0        289.0        0.0  172.0      0.0           0.0   
1    49.0      160.0        180.0        0.0  156.0      1.0           1.0   
2    37.0      130.0        283.0        0.0   98.0      0.0           0.0   
3    48.0      138.0        214.0        0.0  108.0      1.5           1.0   
4    54.0      150.0        195.0        0.0  122.0      0.0           0.0   
..    ...        ...          ...        ...    ...      ...           ...   
913  45.0      110.0        264.0        0.0  132.0      1.2           1.0   
914  68.0      144.0        193.0        1.0  141.0      3.4           1.0   
915  57.0      130.0        131.0        0.0  115.0      1.2           1.0   
916  57.0      130.0        236.0        0.0  174.0      0.0           1.0   
917  38.0      138.0        175.0        0.0  173.0      0.0           0.0   

    Sex ChestPainType RestingECG ExerciseAngina ST_Slope  
0     M           ATA     Normal              N       Up  
1     F           NAP     Normal              N     Flat  
2     M           ATA         ST              N       Up  
3     F           ASY     Normal              Y     Flat  
4     M           NAP     Normal              N       Up  
..   ..           ...        ...            ...      ...  
913   M            TA     Normal              N     Flat  
914   M           ASY     Normal              N     Flat  
915   M           ASY     Normal              Y     Flat  
916   F           ATA        LVH              N     Flat  
917   M           NAP     Normal              N       Up  

[918 rows x 12 columns]

In [5]:
df2=dataset1

df2 = pd.get_dummies(df2, dtype=int,drop_first=True)

In [6]:
df2

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
0    40.0      140.0        289.0        0.0  172.0      0.0           0.0   
1    49.0      160.0        180.0        0.0  156.0      1.0           1.0   
2    37.0      130.0        283.0        0.0   98.0      0.0           0.0   
3    48.0      138.0        214.0        0.0  108.0      1.5           1.0   
4    54.0      150.0        195.0        0.0  122.0      0.0           0.0   
..    ...        ...          ...        ...    ...      ...           ...   
913  45.0      110.0        264.0        0.0  132.0      1.2           1.0   
914  68.0      144.0        193.0        1.0  141.0      3.4           1.0   
915  57.0      130.0        131.0        0.0  115.0      1.2           1.0   
916  57.0      130.0        236.0        0.0  174.0      0.0           1.0   
917  38.0      138.0        175.0        0.0  173.0      0.0           0.0   

     Sex_M  ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  \
0        1                  1                  0                 0   
1        0                  0                  1                 0   
2        1                  1                  0                 0   
3        0                  0                  0                 0   
4        1                  0                  1                 0   
..     ...                ...                ...               ...   
913      1                  0                  0                 1   
914      1                  0                  0                 0   
915      1                  0                  0                 0   
916      0                  1                  0                 0   
917      1                  0                  1                 0   

     RestingECG_Normal  RestingECG_ST  ExerciseAngina_Y  ST_Slope_Flat  \
0                    1              0                 0              0   
1                    1              0                 0              1   
2                    0              1                 0              0   
3                    1              0                 1              1   
4                    1              0                 0              0   
..                 ...            ...               ...            ...   
913                  1              0                 0              1   
914                  1              0                 0              1   
915                  1              0                 1              1   
916                  0              0                 0              1   
917                  1              0                 0              0   

     ST_Slope_Up  
0              1  
1              0  
2              1  
3              0  
4              1  
..           ...  
913            0  
914            0  
915            0  
916            0  
917            1  

[918 rows x 16 columns]

In [7]:
df2.columns

Index(['Age', 'RestingBP', 'Cholesterol', 'FastingBS', 'MaxHR', 'Oldpeak',
       'HeartDisease', 'Sex_M', 'ChestPainType_ATA', 'ChestPainType_NAP',
       'ChestPainType_TA', 'RestingECG_Normal', 'RestingECG_ST',
       'ExerciseAngina_Y', 'ST_Slope_Flat', 'ST_Slope_Up'],
      dtype='object')

In [8]:
indep_X=df2[['Age', 'RestingBP', 'Cholesterol', 'FastingBS', 'MaxHR', 'Oldpeak',
             'Sex_M', 'ChestPainType_ATA', 'ChestPainType_NAP', 'ChestPainType_TA', 
             'RestingECG_Normal', 'RestingECG_ST', 'ExerciseAngina_Y', 'ST_Slope_Flat', 'ST_Slope_Up']]

In [9]:
#indep_X=df2.drop('HeartDisease',inplace=True ,axis=1)


In [10]:
indep_X

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  Sex_M  \
0    40.0      140.0        289.0        0.0  172.0      0.0      1   
1    49.0      160.0        180.0        0.0  156.0      1.0      0   
2    37.0      130.0        283.0        0.0   98.0      0.0      1   
3    48.0      138.0        214.0        0.0  108.0      1.5      0   
4    54.0      150.0        195.0        0.0  122.0      0.0      1   
..    ...        ...          ...        ...    ...      ...    ...   
913  45.0      110.0        264.0        0.0  132.0      1.2      1   
914  68.0      144.0        193.0        1.0  141.0      3.4      1   
915  57.0      130.0        131.0        0.0  115.0      1.2      1   
916  57.0      130.0        236.0        0.0  174.0      0.0      0   
917  38.0      138.0        175.0        0.0  173.0      0.0      1   

     ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  \
0                    1                  0                 0   
1                    0                  1                 0   
2                    1                  0                 0   
3                    0                  0                 0   
4                    0                  1                 0   
..                 ...                ...               ...   
913                  0                  0                 1   
914                  0                  0                 0   
915                  0                  0                 0   
916                  1                  0                 0   
917                  0                  1                 0   

     RestingECG_Normal  RestingECG_ST  ExerciseAngina_Y  ST_Slope_Flat  \
0                    1              0                 0              0   
1                    1              0                 0              1   
2                    0              1                 0              0   
3                    1              0                 1              1   
4                    1              0                 0              0   
..                 ...            ...               ...            ...   
913                  1              0                 0              1   
914                  1              0                 0              1   
915                  1              0                 1              1   
916                  0              0                 0              1   
917                  1              0                 0              0   

     ST_Slope_Up  
0              1  
1              0  
2              1  
3              0  
4              1  
..           ...  
913            0  
914            0  
915            0  
916            0  
917            1  

[918 rows x 15 columns]

In [11]:
dep_Y=df2['HeartDisease']

In [12]:
dep_Y

0      0.0
1      1.0
2      0.0
3      1.0
4      0.0
      ... 
913    1.0
914    1.0
915    1.0
916    1.0
917    0.0
Name: HeartDisease, Length: 918, dtype: float64

In [13]:
indep_X = np.abs(indep_X)

In [59]:
kbest=selectkbest(indep_X,dep_Y,3)    

In [61]:

acclog=[]
accsvml=[]
accsvmnl=[]
accknn=[]
accnav=[]
accdes=[]
accrf=[]



In [63]:

X_train, X_test, y_train, y_test=split_scalar(kbest,dep_Y)   


In [65]:
        
classifier,Accuracy,report,X_test,y_test,cm=logistic(X_train,y_train,X_test)
acclog.append(Accuracy)

classifier,Accuracy,report,X_test,y_test,cm=svm_linear(X_train,y_train,X_test)  
accsvml.append(Accuracy)
    
classifier,Accuracy,report,X_test,y_test,cm=svm_NL(X_train,y_train,X_test)  
accsvmnl.append(Accuracy)
    
classifier,Accuracy,report,X_test,y_test,cm=knn(X_train,y_train,X_test)  
accknn.append(Accuracy)
    
classifier,Accuracy,report,X_test,y_test,cm=Navie(X_train,y_train,X_test)  
accnav.append(Accuracy)
    
classifier,Accuracy,report,X_test,y_test,cm=Decision(X_train,y_train,X_test)  
accdes.append(Accuracy)
    
classifier,Accuracy,report,X_test,y_test,cm=random(X_train,y_train,X_test)  
accrf.append(Accuracy)
    
result=selectk_Classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf)


C:\Users\LENOVO\AppData\Local\Temp\ipykernel_3304\1247720774.py:96: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  dataframe['Logistic'][idex]=acclog[number]
C:\Users\LENOVO\AppData\Local\Temp\ipykernel_3304\1247720774.py:97: FutureWarning: C

In [37]:
#k=5
result

Logistic     SVMl     SVMnl       KNN     Navie  Decision    Random
ChiSquare  0.808696  0.76087  0.821739  0.817391  0.808696  0.769565  0.795652

In [47]:
#k=6
result

Logistic     SVMl     SVMnl       KNN     Navie  Decision    Random
ChiSquare  0.786957  0.76087  0.830435  0.821739  0.804348  0.782609  0.795652

In [57]:
#k=4
result

Logistic     SVMl     SVMnl       KNN     Navie  Decision    Random
ChiSquare  0.795652  0.76087  0.817391  0.808696  0.821739  0.791304  0.786957

In [67]:
#k=3
result

Logistic     SVMl     SVMnl       KNN     Navie  Decision    Random
ChiSquare  0.769565  0.76087  0.821739  0.782609  0.808696  0.808696  0.804348

# k=6, SVMnl is the best Model which has 83% Accuracy